### PR TITLE
Share heap usage numbers to support cache cleanup in VS code

### PR DESCRIPTION
**Pull request by rchiodo**
VS code uses pointer compression which seems to share the heap with all the worker threads. This means the `getHeapStatistics` function is not returning the correct amount of memory in use for a thread.

This change shares this number with the other threads so they can take total usage into account when deciding whether or not to clean up the heap.

### CODE REVIEW

File: packages/pyright-internal/src/analyzer/cacheManager.ts [hunk 1]
```diff
@@ -8,6 +8,9 @@
  * if memory usage approaches the max heap space.
  */
 
+import { HeapInfo } from 'v8';
```
Comment (erictraut) [inline packages/pyright-internal/src/analyzer/cacheManager.ts:11]: Will this break browser support for vscode.dev?

File: packages/pyright-internal/src/analyzer/cacheManager.ts [hunk 2]
```diff
@@ -24,12 +27,41 @@ export interface CacheOwner {
...
+                view[index] = 0;
+            });
+        }
+    }
+
+    handleCachedUsageBufferMessage(msg: AnalysisRequest) {
+        if (msg.requestType === 'cacheUsageBuffer') {
+            const index = parseInt(msg.data || '0');
+            const buffer = msg.sharedUsageBuffer;
+            if (buffer && index) {
```
Comment (erictraut) [inline packages/pyright-internal/src/analyzer/cacheManager.ts:58]: Isn't 0 a legal index? Or is an index of 0 reserved for the main thread?

This `if` statement will evaluate to falsy if index is 0.

File: packages/pyright-internal/src/backgroundAnalysis.ts [hunk 3]
```diff
@@ -19,18 +19,22 @@ import { ServiceProvider } from './common/serviceProvider';
 import { getRootUri } from './common/uri/uriUtils';
 
 export class BackgroundAnalysis extends BackgroundAnalysisBase {
-    constructor(serviceProvider: ServiceProvider) {
+    constructor(serviceProvider: ServiceProvider, workerIndex = BackgroundThreadIndex) {
```
Comment (erictraut) [inline packages/pyright-internal/src/backgroundAnalysis.ts:22]: This seems fragile. If more than one BackgroundAnalysis object is instantiated and the caller forgets to pass in a unique index, the two will now step on each other's data in the shared buffer. Is there a way to make this more robust, like dynamically generating a unique ID or index during construction of the BackgroundAnalysis?

**Review by erictraut — APPROVED** [review]
Looks good to me. Just a couple of questions to consider. No need to respond or make changes if you have good answers for them.

File: packages/pyright-internal/src/analyzer/cacheManager.ts (hunk 1 above)
Comment (rchiodo) [reply packages/pyright-internal/src/analyzer/cacheManager.ts:11]: I don't know yet.  I was going to handle this when we port this to Pylance. If that's the case, I can just import the type. Actually I'll just do that now.

File: packages/pyright-internal/src/analyzer/cacheManager.ts (hunk 2 above)
Comment (rchiodo) [reply packages/pyright-internal/src/analyzer/cacheManager.ts:58]: 0 is the index reserved for the main thread. I'll add a comment.

File: packages/pyright-internal/src/backgroundAnalysis.ts (hunk 3 above)
Comment (rchiodo) [reply packages/pyright-internal/src/backgroundAnalysis.ts:22]: Oh good point. I'll generate it from a static. This is always called on the main thread.

**Pull request merged**